xmtp_mls/src/subscriptions/stream_conversations.rs

```diff
@@ -131,6 +131,86 @@ where
     }
 }
 
+pin_project! {
+    /// A stream that polls two inner streams.
+    /// this stream closes once either one of the streams close.
+    /// gives priority to S1 (S1 will always be polled first.)
+    /// If S1 finishes before S2, will attempt to drain S2 of items.
+    pub struct MultiplexedStream<S1, S2, T> {
+        #[pin] s1: S1,
+        #[pin] s2: S2,
+        queue: VecDeque<T>,
+        terminated: bool,
+    }
+}
+
+impl<S1, S2, T> MultiplexedStream<S1, S2, T> {
+    pub fn new(s1: S1, s2: S2) -> Self {
+        Self {
+            s1,
+            s2,
+            queue: Default::default(),
+            terminated: false,
+        }
+    }
+}
+
+// TODO: unit test this
+impl<S1, S2, T> Stream for MultiplexedStream<S1, S2, T>
+where
+    S1: Stream<Item = T>,
+    S2: Stream<Item = T>,
+{
+    type Item = T;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut this = self.as_mut().project();
+        if *this.terminated {
```

Comment:
there's definitely a bug here where the FusedStream implementation will return terminated as true, while its still possible that the stream has items sitting in `queue`